src/Microsoft.AspNetCore.SystemWebAdapters.CoreServices/HttpApplication/HttpApplicationExtensions.cs

```diff
@@ -0,0 +1,157 @@
+{
+    public static ISystemWebAdapterBuilder AddHttpApplication(this ISystemWebAdapterBuilder builder, Action<HttpApplicationOptions> configure)
+    {
+        ArgumentNullException.ThrowIfNull(builder);
+
+        builder.Services.TryAddSingleton<HttpApplicationPooledObjectPolicy>();
+        builder.Services.AddTransient<IStartupFilter>(ctx => ctx.GetRequiredService<HttpApplicationPooledObjectPolicy>());
+        builder.Services.TryAddSingleton<IPooledObjectPolicy<HttpApplication>>(ctx => ctx.GetRequiredService<HttpApplicationPooledObjectPolicy>());
+        builder.Services.TryAddSingleton<ObjectPool<HttpApplication>>(sp =>
+        {
+            var options = sp.GetRequiredService<IOptions<HttpApplicationOptions>>();
+            var policy = sp.GetRequiredService<IPooledObjectPolicy<HttpApplication>>();
+
+            var provider = new DefaultObjectPoolProvider
+            {
+                MaximumRetained = options.Value.PoolSize,
+            };
+
+            return provider.Create(policy);
+        });
+
+        builder.Services.AddOptions<HttpApplicationOptions>()
+            .Configure(configure)
+            .PostConfigure(c => c.MakeReadOnly());
+
+        return builder;
+    }
+
+    public static ISystemWebAdapterBuilder AddHttpApplication<TApp>(this ISystemWebAdapterBuilder builder, Action<HttpApplicationOptions> configure)
+        where TApp : HttpApplication
+    {
+        ArgumentNullException.ThrowIfNull(builder);
+        ArgumentNullException.ThrowIfNull(configure);
+
+        builder.AddHttpApplication(options =>
+        {
+            options.ApplicationType = typeof(TApp);
+
+            configure(options);
+        });
+
+        return builder;
+    }
+
+    internal static void UseHttpApplication(this IApplicationBuilder app)
+    {
+        if (app.AreHttpApplicationEventsRequired())
+        {
+            app.UseMiddleware<HttpApplicationMiddleware>();
+            app.UseHttpApplicationEvent(ApplicationEvent.BeginRequest);
+        }
+    }
+
+    internal static void UseHttpApplicationEvent(this IApplicationBuilder app, params ApplicationEvent[] preEvents)
+        => app.UseHttpApplicationEvent(preEvents, Array.Empty<ApplicationEvent>());
+
+    internal static void UseHttpApplicationEvent(this IApplicationBuilder app, ApplicationEvent[] preEvents, ApplicationEvent[] postEvents)
+    {
+        if (app.AreHttpApplicationEventsRequired())
+        {
+            app.Use(async (ctx, next) =>
+            {
+                var appFeature = ctx.Features.GetRequired<IHttpApplicationFeature>();
+                var endFeature = ctx.Features.GetRequired<IHttpResponseEndFeature>();
+
+                foreach (var @event in preEvents)
+                {
+                    await appFeature.RaiseEventAsync(@event);
+
+                    if (endFeature.IsEnded)
+                    {
+                        return;
+                    }
+                }
+
+                await next(ctx);
```

Comment:
should the postEvents still run if an exception is thrown?

Reply:
The answer is "What does ASP.NET do?"

Reply:
and the answer is "no" for these, but "yes" to the events after logging. I've added some exception handling in to take care of these (and expanded the tests to cover throwing in each of the events)